src/reflect/scala/reflect/internal/transform/Erasure.scala

```diff
@@ -246,39 +242,50 @@ trait Erasure {
    *     parents |Ps|, but with duplicate references of Object removed.
    *   - for all other types, the type itself (with any sub-components erased)
    */
-  def erasure(sym: Symbol): ErasureMap =
-    if (sym == NoSymbol || !sym.enclClass.isJavaDefined) scalaErasure
-    else if (verifyJavaErasure && sym.isMethod) verifiedJavaErasure
-    else javaErasure
+  def erasure(sym: Symbol): ErasureMap = {
+    if (sym == NoSymbol) return scalaErasure
+    val enclosing = sym.enclClass
+    if (enclosing.isJavaDefined) {
+      if (verifyJavaErasure && sym.isMethod) verifiedJavaErasure
+      else javaErasure
+    }
+    else if (enclosing.isScala3Defined) scala3Erasure
+    else scalaErasure
+  }
 
   /** This is used as the Scala erasure during the erasure phase itself
    *  It differs from normal erasure in that value classes are erased to ErasedValueTypes which
    *  are then later converted to the underlying parameter type in phase posterasure.
+   *
+   * @param symOfTp used to determine the erasure mode for the type,
+   *        e.g. in `SymbolPair#highErased`, `sym` may be an anonymous class for a SAM type,
+   *        but `symOfTp` may be the a bridge method for the SAM method being erased.
    */
-  def specialErasure(sym: Symbol)(tp: Type): Type =
+  def specialErasure(sym: Symbol)(tp: Type, symOfTp: Symbol): Type =
```

Comment:
It seems the only callsites where `sym != symOfTp` are the ones in SymbolPairs.

Maybe it would be correct to change those instead.

```diff
-    def lowErased: Type     = erasure.specialErasure(base)(low.tpe)
+    def lowErased: Type     = erasure.specialErasure(low)(low.tpe)
```

It goes back all the way here: https://github.com/scala/scala/commit/44b9cf0ca9#diff-271d576b3a80359cbc5db9e1a8dd2629c2cb1d99bf8a30854c64742f98c7a6b0R964

Let's at least try that out and see if anything breaks.

Reply:
I'm trying this now

Reply:
seems to have worked

Reply:
great!